#11749 Attempt to fix UTF8 filenames

burzum opened the pull request:
See #11749

burzum committed:
#11749 Attempt to fix UTF8 filenames

saeideng commented on src/Filesystem/File.php:
```diff
@@ -82,9 +83,10 @@ class File
+        $this->Folder = new Folder($splInfo->getPath(), $create, $mode);
         if (!is_dir($path)) {
-            $this->name = basename($path);
+            $this->name = $splInfo->getBasename();
```
this method does not work on browser ( apache ) 
but console is OK 
```php
<?php
$path='tests/permissions/نام فارسی.php';
$splInfo = new \SplFileInfo($path);

var_dump($splInfo->getFilename());
var_dump($splInfo->getBasename());
var_dump(basename($path));
```

burzum committed:
#11749 Changing the way the filename is read

* https://stackoverflow.com/questions/32115609/basename-fail-when-file-name-start-by-an-accent/40891025#40891025
* https://bugs.php.net/bug.php?id=62119

saeideng committed:
add a function to simulate the php basename()

stickler-ci committed:
Fixing style errors.

saeideng committed:
add doc for _basename()

saeideng committed:
add some test for _basename()

saeideng committed:
add more test to _basename

test for "rtrim($name, $ext) === ''" section

burzum commented:
Looks good now!

markstory commented on tests/TestCase/Filesystem/FileTest.php:
```diff
@@ -123,6 +123,69 @@ function_exists('mime_content_type') &&
+        if ($suffix === null) {
+            $this->assertEquals(basename($path), $method->invokeArgs(null, [$path]));
+        } else {
+            $this->assertEquals(basename($path, $suffix), $method->invokeArgs(null, [$path, $suffix]));
```
If you want to call this method why is it not public? Coupling tests to 'implementation details' makes tests harder to maintain as private features are now public.

saeideng replied:
we can add/move these tests to other public tests
i just want to know `_basename()` method is equal to `basename()` 
And i no want to have public method at this time

markstory replied:
Can we use a public interface of the class then?

saeideng replied:
>Can we use a public interface of the class then?

1)having extended class withing testCase and having a method for wrapping _basename() 
like the https://github.com/cakephp/cakephp/blob/master/tests/TestCase/Mailer/EmailTest.php#L89-L92 in to testcase?

2) or moving test to public methods?

in case 2 , not fully - we can not cover all case - i'm not sure

markstory replied:
Can we not create empty files for each case and check their names?

saeideng replied:
I'll move these tests  to `name()` without having real file

markstory replied:
Could the behavior of `_basename` be tested based on how file creation and `name()` are working?

saeideng commented on src/Filesystem/File.php:
```diff
@@ -368,7 +388,7 @@ public function safe($name = null, $ext = null)
         }
 
-        return preg_replace("/(?:[^\w\.-]+)/", '_', basename($name, $ext));
+        return preg_replace("/(?:[^\w\.-]+)/", '_', static::_basename($name, $ext));
```
probably needs a bit test to testing this method 
`public function safe($name = null, $ext = null)`

saeideng commented on src/Filesystem/File.php:
```diff
@@ -352,6 +354,24 @@ public function name()
+    {
+        $splInfo = new SplFileInfo($path);
+        $name = ltrim($splInfo->getFilename(), DS);
+        if ($ext === null || rtrim($name, $ext) === '') {
```
`rtrim()`  must change with proper method
now it returns invalid result

saeideng committed:
add proper method to cut the string

stickler-ci committed:
Fixing style errors.

saeideng commented:
- there is problem with `/نام.ext` on Windowns yet , Linux is OK
also , `folderName/نام.ext` is OK in both 
https://ci.appveyor.com/project/cakephp/cakephp/build/1.0.9829#L274

- I added `(mb_strlen($path) === strlen($path))` for using `basename()` if file name/path was not `multibyte`

- I will move test from `_basename()` to `name()` if I able

saeideng committed:
triming DS is not  enough for `Spl::getFilename()`

`SplFileInfo->getFilename('/filename.ext')` returns `/filename.ext` in both linux and windows

saeideng commented:
>looks there is problem with /نام.ext on Windowns yet , Linux is OK
also , folderName/نام.ext is OK in both
https://ci.appveyor.com/project/cakephp/cakephp/build/1.0.9829#L274

Tests passed on windows (AppVeyor)

saeideng committed:
move tests from protected method  to name()

saeideng commented:
finished
